### PR TITLE
263.2 - Support internal hash links (markdown related)

### PR DESCRIPTION
**Pull request by W3stside**
Merges into #390 

Adds capabilities for internal links with markdown

## Typical over-engineered react way of doing this

Adds `ScrollToTop` comp for easier access to links

### CODE REVIEW

File: src/custom/components/ScrollToTop/index.tsx [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    &:hover {
+        opacity: 1;
+    }
+
+    > svg {
+        margin: 0 0.4rem;
+    }
+
+    transition: opacity 0.4s ease-in-out;
```
Comment (biocom) [inline src/custom/components/ScrollToTop/index.tsx:43]: Nit: why not move it right after line 33? The way I look at the code is in this order:
- Properties for this current element
- Any pseudo elements (&:hover, etc.) still part of this property.
- Target any children

**Comment by biocom** [comment]
Thank you. My only comment is, I wouldn't overlay the Scroll to top button on the content itself. Ideally outside of the content container. Also I don't see the direct need/utility for desktop, but perhaps more on mobile (which doesn't seem to be optimized for it yet).

File: src/custom/components/ScrollToTop/index.tsx (hunk 1 above)
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:43]: :shrug: i can do that

**Comment by W3stside** [comment]
> Thank you. My only comment is, I wouldn't overlay the Scroll to top button on the content itself. Ideally outside of the content container. Also I don't see the direct need/utility for desktop, but perhaps more on mobile (which doesn't seem to be optimized for it yet).

ok i can give that a go - tried it first and didn't really like it but can go again. maybe it is better only on mobile as well, will give that a run

**Review by alfetopito — COMMENTED** [review]
Is this supposed to be visible in the page yet?
Cause I did not find it anywhere

**Comment by W3stside** [comment]
> Is this supposed to be visible in the page yet?
> Cause I did not find it anywhere

What changes are you referring to?

**Comment by alfetopito** [comment]
> > Is this supposed to be visible in the page yet?
> > Cause I did not find it anywhere
> 
> What changes are you referring to?

The privacy policy and whatever markdown you are setting up

**Comment by W3stside** [comment]
> > > Is this supposed to be visible in the page yet?
> > > Cause I did not find it anywhere
> > 
> > 
> > What changes are you referring to?
> 
> The privacy policy and whatever markdown you are setting up

/#/privacy-policy or /#/cookie-policy

**Review by Velenir — COMMENTED** [review]
Hey guys! This PR caught my eye, because _markdown_ ✔️ ⬇️ 

Here are a few comments, if you don't mind :wink:

**Comment by biocom** [comment]
@Velenir Is back!!!!??!?! Thanks for sharing :D

**Comment by biocom** [comment]
On that note, perhaps we could prevent using any listeners, as @Velenir pointed out those are 'expensive' performance wise. The first thing I check into are the IntersectionObservers, where we could detect if a certain element it out of view. Example: https://m.signalvnoise.com/how-to-back-to-top-button-without-scroll-events/

File: src/custom/components/ScrollToTop/index.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  useEffect(() => {
+    const scrollListener = () => debouncedSetPosition(window.scrollY)
+
+    window.addEventListener('scroll', scrollListener)
+
+    return () => window.removeEventListener('scroll', scrollListener)
+  }, [debouncedSetPosition])
+
+  if (position < 250) return null
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:60]: no i didnt want it to be present after scrolling +1px, so i set a random magic 250 to give it some space before appearing

File: src/custom/components/Markdown/index.tsx [hunk 3]
```diff
@@ -1,30 +1,38 @@
...
+import ScrollToTop from '../ScrollToTop'
 
 interface MarkdownParams {
   content: string
   title?: ReactNode
 }
 
+const CustomReactMarkdown = (props: ReactMarkdownPropsBase & { children: string }) => (
+  <ReactMarkdown {...props} renderers={{ heading: HeadingRenderer, link: LinkRenderer }} allowDangerousHtml>
+    {props.children}
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/Markdown/index.tsx:16]: :+1:

File: src/custom/components/ScrollToTop/index.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    > svg {
+        margin: 0 0.4rem;
+    }
+
+    transition: opacity 0.4s ease-in-out;
+`
+
+export default function ScrollToTop(props: { styleProps?: WrapperProps }) {
+  const [position, setPosition] = useState(window.scrollY)
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:49]: :+1:

File: src/custom/components/ScrollToTop/index.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  const [, debouncedSetPosition] = useDebouncedChangeHandler(position, setPosition, 500)
+
+  const handleClick = () => window.scrollTo(0, 0)
+
+  useEffect(() => {
+    const scrollListener = () => debouncedSetPosition(window.scrollY)
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:55]: nice! :+1:

File: src/custom/components/ScrollToTop/index.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    window.addEventListener('scroll', scrollListener)
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:57]: yess! come back @Velenir ;)

File: src/custom/components/Markdown/renderers.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * HeadingRenderer is a custom renderer
+ * It parses the heading and attaches an id to it to be used as an anchor
+ */
+const HeadingRenderer = (props: { level: number; children: ReactNode }) => {
+  const children = React.Children.toArray(props.children)
+  const text = children.reduce(flatten, '')
+  const slug = text.toLowerCase().replace(/\W/g, '-')
+  return React.createElement('h' + props.level, { id: slug }, props.children)
+}
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/Markdown/renderers.tsx:23]: great suggestion!

File: src/custom/components/ScrollToTop/index.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:49]: :+1: yep, much more optimised, thx

File: src/custom/components/Markdown/renderers.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+// Another overcomplicated React thing? ok!
+// react-markdown doesn't auto add ids to header tags
+// making TOCs impossible without some annoying plugins to install etc
+// https://github.com/remarkjs/react-markdown/issues/404
+function flatten(text: string, child: ReactNode): any {
+  return typeof child === 'string'
+    ? text + child
+    : React.Children.toArray((child as ReactElement).props.children).reduce(flatten, text)
+}
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/Markdown/renderers.tsx:12]: yep good idea, thx

File: src/custom/components/HashLink/index.tsx [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+  const location = useLocation()
+
+  useEffect(() => {
+    const id = location.hash.slice(1)
+    const elem = document.getElementById(id)
+
+    if (!elem) return
+
+    window.scrollTo(0, elem.offsetTop)
+  }, [location.hash])
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/HashLink/index.tsx]: made changes in the hash link checking `props.to` against `location.hash` - this should optimise no?

File: src/custom/components/HashLink/index.tsx [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+const HashLink = (props: LinkProps) => {
+  const location = useLocation()
+
+  useEffect(() => {
+    const id = location.hash.slice(1)
+    const elem = document.getElementById(id)
+
+    if (!elem) return
+
+    window.scrollTo(0, elem.offsetTop)
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/HashLink/index.tsx]: great call!

**Review by alfetopito — CHANGES_REQUESTED** [review]
On https://pr398--gpswapui.review.gnosisdev.com/#/cookie-policy page, the list to Privacy policy adds an extra `#`:

screenshot_2021-04-12_10-18-51

leading to https://pr398--gpswapui.review.gnosisdev.com/#/swap#/privacy-policy

**Review by anxolin — COMMENTED** [review]
APPROVED.

Just delete the dead code. 

It feels unlikely that this will be an issue, but note that you are injecting the scroll to top on each markdown. Probably the scroll to top should be sth we guarantee there's only one instance for the whole app.  No need to change anything right now, cause afaik we only will use one MD compoent per content page

**Comment by W3stside** [comment]
> APPROVED.
> 
> Just delete the dead code.
> 
> It feels unlikely that this will be an issue, but note that you are injecting the scroll to top on each markdown. Probably the scroll to top should be sth we guarantee there's only one instance for the whole app. No need to change anything right now, cause afaik we only will use one MD compoent per content page

good point, will re-explore

File: src/custom/components/ScrollToTop/index.tsx [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+interface WrapperProps {
+  top?: string
+  right?: string
+  bottom?: string
+  left?: string
+  background?: string
+}
+
+const Wrapper = styled.div<WrapperProps>`
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:14]: removed scrollToTop

File: src/custom/components/ScrollToTop/index.tsx [hunk 12]
```diff
@@ -0,0 +1,74 @@
+import React, { useEffect, useState } from 'react'
```
[the remark this answers is not in the document]
Comment (W3stside) [reply src/custom/components/ScrollToTop/index.tsx:1]: removed

**Review by anxolin — APPROVED** [review]

**Review by alfetopito — APPROVED** [review]

**Review by biocom — APPROVED** [review]
Looks good. Will tackle some text styling issues in a separate PR.

I think the reading format with the narrow container is perfect as it reads easier than long sentences. Curious what user feedback will tell.

**Pull request merged**